### PR TITLE
fix(OwnerReferenceUpdates): Handle error logging for owner reference updates

### LINKED ISSUE #3776: Race observed when patching the envoy bootstrap secret in replicated control plane environment
**Bug description**:
There is a bug where multiple instances of osm-controller attempt to update the same k8s secret corresponding to Envoy's bootstrap certificate.

The code in https://github.com/openservicemesh/osm/blob/main/pkg/k8s/announcement_handlers.go#L18 errors when multiple replicated osm-controller instances attempt to update the k8s secret on Pod Add events.
```
{"level":"error","component":"kube-controller","error":"Operation cannot be fulfilled on secrets \"envoy-bootstrap-config-431aca64-705f-4b80-85c6-7096c4c4e678\": the object has been modified; please apply your changes to the latest version and try again","time":"2021-07-16T19:47:26Z","file":"announcement_handlers.go:61","message":"Failed to update OwnerReference for Secret bookwarehouse/envoy-bootstrap-config-431aca64-705f-4b80-85c6-7096c4c4e678 to reference Pod bookwarehouse/bookwarehouse-79c6f8fdc4-gpgtc"}
```
```
{"level":"error","component":"kube-controller","error":"Operation cannot be fulfilled on secrets \"envoy-bootstrap-config-b23262a9-bc64-4ec6-8357-4b31794abc2c\": the object has been modified; please apply your changes to the latest version and try again","time":"2021-07-16T19:46:36Z","file":"announcement_handlers.go:61","message":"Failed to update OwnerReference for Secret bookbuyer/envoy-bootstrap-config-b23262a9-bc64-4ec6-8357-4b31794abc2c to reference Pod bookbuyer/bookbuyer-7b68f98fd4-dl46c"}
{"level":"error","component":"kube-controller","error":"Operation cannot be fulfilled on secrets \"envoy-bootstrap-config-85fbdbae-f231-4e22-8e25-401c8e7dac0e\": the object has been modified; please apply your changes to the latest version and try again","time":"2021-07-16T19:46:55Z","file":"announcement_handlers.go:61","message":"Failed to update OwnerReference for Secret bookthief/envoy-bootstrap-config-85fbdbae-f231-4e22-8e25-401c8e7dac0e to reference Pod bookthief/bookthief-5d88db99ff-2g9zz"}
{"level":"error","component":"kube-controller","error":"Operation cannot be fulfilled on secrets \"envoy-bootstrap-config-365851ab-20d6-48e2-bc6a-12ed8343d6a2\": the object has been modified; please apply your changes to the latest version and try again","time":"2021-07-16T19:47:10Z","file":"announcement_handlers.go:61","message":"Failed to update OwnerReference for Secret bookstore/envoy-bootstrap-config-365851ab-20d6-48e2-bc6a-12ed8343d6a2 to reference Pod bookstore/bookstore-5487c8bd9c-6mwhv"}
{"level":"error","component":"kube-controller","error":"Operation cannot be fulfilled on secrets \"envoy-bootstrap-config-d9e2bab0-8f53-47dc-a3fc-87af0eb1c4d9\": the object has been modified; please apply your changes to the latest version and try again","time":"2021-07-16T20:30:05Z","file":"announcement_handlers.go:61","message":"Failed to update OwnerReference for Secret bookstore/envoy-bootstrap-config-d9e2bab0-8f53-47dc-a3fc-87af0eb1c4d9 to reference Pod bookstore/bookstore-v2-5f8b4785cc-4s7fr"}
```

**Affected area (please mark with X where applicable)**:
- Control Plane        [X]

**Expected behavior**:
Patching of the k8s secret must happen atomically in replicated control plane environments.

**Steps to reproduce the bug (as precisely as possible)**:
1. `osm install --set OpenServiceMesh.osmController.replicaCount=2`
2. Run the manual demo described in https://docs.openservicemesh.io/docs/getting_started/manual_demo/

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:

Check for conflicts before logging an error for failed updates of owner
reference to the k8s secrets. This is necessary when multiple instances
of osm-controller are deployed, as they will all try to update the same
resource and only one of them will update it, which is sufficient.

Fixes #3776

Signed-off-by: Sneha Chhabria <snchh@microsoft.com>

**Testing done**: Manually verified that this error no longer occurs with multiple osm-controllers 

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Demo                       | [ ] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [ ] |
| Upgrade                    | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? `no` 
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change? `no`

### CODE REVIEW

**Review by jaellio — APPROVED** [review]
LGTM!

File: pkg/k8s/announcement_handlers.go [hunk 1]
```diff
@@ -58,7 +59,11 @@ func PatchSecretHandler(kubeClient kubernetes.Interface) chan struct{} {
 				})
 
 				if _, err = kubeClient.CoreV1().Secrets(namespace).Update(context.Background(), secret, metav1.UpdateOptions{}); err != nil {
-					log.Error().Err(err).Msgf("Failed to update OwnerReference for Secret %s/%s to reference Pod %s/%s", namespace, secretName, namespace, podName)
+					// There might be conflicts when multiple controllers try to update the same resource
+					// One of the controllers will successfully update the resource, hence conflicts shoud be ignored and not treated as an error
+					if apierrors.IsConflict(err) {
```
Comment (shashankram) [inline pkg/k8s/announcement_handlers.go:64]: I assume you intended to log the error only when there's an error that's not a conflict error?
```suggestion
					if !apierrors.IsConflict(err) {
```

**Review by shashankram — CHANGES_REQUESTED** [review]

File: pkg/k8s/announcement_handlers.go (hunk 1 above)
Comment (snehachhabria) [reply pkg/k8s/announcement_handlers.go]: I had a local change that I forgot to push

**Review by shashankram — APPROVED** [review]
Looks good now.

**Pull request merged**